File: codeForces/D_Sakurako_s_Hobby.cpp
#include<bits/stdc++.h>
using namespace std;

#ifndef ONLINE_JUDGE
    #include "algoDebug.h"
#else
    #define debug(x...)
    #define RAYA 
    #define raya 
#endif

void sol () {
    int n;
    cin >> n;

    string s;
    vector<int> p(n + 1);
    for (int i = 1; i <= n; i++) {
        cin >> p[i];
    }
    cin >> s;

    int cnt = 1;
    vector<int> color(n + 1);
    vector<int> res(n + 1, 0);
    for (int i = 1; i <= n; i++) {
        if (color[i]) continue;

        for (int curr = i; !color[curr]; curr = p[curr]) {
            color[curr] = cnt;
            if (s[curr - 1] == '0') {
                res[cnt]++;
            }
        }

        cnt++;
    }

    for (int i = 1; i <= n; i++) {
        int who = color[i];
        cout << max(0, res[who]) << " \n"[i == n];
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    int t = 1;
    cin >> t;

    while(t--) {
        sol();
    }

    return 0;
}

File: codeForces/D_Sakurako_s_Hobby.py
import sys
input = sys.stdin.read
data = input().split()

index = 0
t = int(data[index])
index += 1

for _ in range(t):
    n = int(data[index])
    index += 1
    
    p = [0] * (n + 1)
    for i in range(1, n + 1):
        p[i] = int(data[index])
        index += 1
    
    s = data[index]
    index += 1

    cnt = 1
    color = [0] * (n + 1)
    res = [0] * (n + 1)
    
    for i in range(1, n + 1):
        if color[i]:
            continue

        curr = i
        while not color[curr]:
            color[curr] = cnt
            if s[curr - 1] == '0':
                res[cnt] += 1
            curr = p[curr]

        cnt += 1

    # print(f"color: {color[1:]}")
    # print(f"res: {res[1:]}")

    output = []
    for i in range(1, n + 1):
        who = color[i]
        print(res[who], end = ' ')
    
    print()